### ISSUE TITLE
Setting state

### ISSUE THREAD

**Issue by silvenon** [Munter/MochaJsdomReact#1]
Not really sure what you're doing [here](https://github.com/Munter/MochaJsdomReact/blob/master/test/test-1.js#L36). If you delete that, tests pass.

When you're doing `isVisible={false}`, you're setting a **property**, state is different. Even if done correctly, your component would end up having `this.props.isVisible` set to `false` and `this.state.isVisible` set to `true`.

If you wanted to set the initial state, the correct way of doing it would be using `getInitialState`:

``` js
  Leaker = React.createClass({
      getInitialState() {
        return {
          isVisible: true
        };
      },

      render() {
        return (
          <div>
            When this HTML innerText is here, the tests fail
            {this.state ? 'foo' : null}
          </div>
        )
      }
    });
```

I'm happy to explain further, but I'm not sure what you were trying to do. What was the purpose of `isVisible`?

**Comment by Munter** [comment]
There is no specific purpose of `isVisible`. The `setState` call combined with the ternary operator in the render function is just a mechanism to trigger a new render cycle and vdom difference. This has gone through a lot of reduction from a massive code base, so there are some things that could probably still be reduced.

**Comment by Munter** [comment]
I removed the unused prop to make the example less confusing

**Comment by silvenon** [comment]
Have you seen [this](https://github.com/jesstelford/react-testing-mocha-jsdom)? I think React should be tested with their [TestUtils](http://facebook.github.io/react/docs/test-utils.html).

**Comment by Munter** [comment]
That article pretty much describes what we are doing, except we created [jsdomify](https://github.com/podio/jsdomify) to encapsulate the functionality of creating a document and leaking global window and document variables.

TestUtils are just convenience. In our full tests we use `TestUtils.renderIntoDocument()`, but for querying elements out of the rendered component we use [unexpected-dom](https://githubcom/Munter/unexoected-dom). I just wanted to remove those possible sources of errors in this minimal example

**Comment by silvenon** [comment]
That way of re-rendering the component is obviously illegal, it should be done within the component. That's the most I could get out of the example.

**Comment by Munter** [comment]
Calling `setState` is illegal?

**Comment by silvenon** [comment]
No, I thought it being used **there** is illegal, but I was wrong, the same thing happens if you use it in `componentDidMount` (but not in `componentWillMount`), which should be fine.

But that's not the issue here, your question was why doesn't the first test fail then and why does the error seem to leak across tests. My bad, I'll play with it more…

**Comment by silvenon** [comment]
[Found it!](http://stackoverflow.com/a/30544088/1247274)

``` js
  before(function () {
    // remove react from the require cache
    for (var key in require.cache) {
      if (key.match(/\/node_modules\/react\//)) {
        delete require.cache[key];
      }
    }

    // ...
  });
```

**Comment by Munter** [comment]
Thanks. We also came to the same conclusion here: https://github.com/facebook/react/issues/4025

**Comment by nikolaymatrosov** [comment]
I want to thank you for making this sandbox repo. It was very helpful to me when I was trying to solve same problem. Additionally I have found out that if you load component from another file you need to delete it from cache too. It isn't necessary when I run my test using only mocha. But it becomes necessary if I try to run them with Wallaby.js.

``` javascript
before(function () {

    for (var key in require.cache) {
      if (key.match(/react/ || key.match(/Leaker/))) {
        delete require.cache[key];
      }
    }
    React = require('react');

    Leaker = require('../src/Leaker');
});
```
